File: legacy/tools/scopecheck.py
#!/usr/bin/env python3
"""Find identifiers used in a component's JS body that are never declared.

`validate.js` catches undefined *calls*; this catches undefined *variables* —
the failure that produced "Can't find variable: editing" and a blank tab. It
inspects only the JS above each component's `return (`, which is where hooks and
derived values live, so prose inside JSX can't create false positives.

SCOPE, deliberately: components only, not the engine functions. That is not an
oversight. An undeclared READ throws a ReferenceError the moment the line runs,
in strict mode or out of it, and the engines are exercised by golden,
protocols and run_all on every build — so an undeclared name there fails the
gate at runtime rather than needing a parser to spot it. Components are the
place a path can go unexercised, because nothing renders them.

Verified by probe: an undeclared read inserted into assessCalcium is caught by
golden, protocols and run_all. This checker does not see it and does not need
to.
"""
import re, sys

raw = open(sys.argv[1]).read()

def strip_noise(t):
    """Remove comments and string literals so prose and class names can't be
    mistaken for identifiers."""
    t = re.sub(r'/\*[\s\S]*?\*/', ' ', t)
    t = re.sub(r'(?m)^\s*//[^\n]*$', ' ', t)
    # Trailing comments too — "// 0 counting, 1 counted" was being parsed as
    # three undeclared identifiers.
    t = re.sub(r'(?m)//[^\n]*$', ' ', t)
    t = re.sub(r'"(?:\\.|[^"\\])*"', '""', t)
    t = re.sub(r"'(?:\\.|[^'\\])*'", "''", t)
    t = re.sub(r'`(?:\\.|[^`\\])*`', '``', t)
    return t

src = raw
# `async function foo` is still a top-level declaration; without the optional
# async the scan missed loadKey and requestPersistence entirely.
toplevel = set(re.findall(r'^(?:async\s+)?(?:function|const|let|var|class)\s+(\w+)', src, re.M))
builtins = {
 'React','Math','Object','Array','Number','String','JSON','Date','Set','Map','Promise','Error',
 'console','window','document','navigator','fetch','setTimeout','clearTimeout','setInterval',
 'clearInterval','requestAnimationFrame','cancelAnimationFrame','parseInt','parseFloat','isNaN',
 'isFinite','useState','useEffect','useMemo','useRef','Blob','FileReader','URL','encodeURIComponent',
 'decodeURIComponent','localStorage','sessionStorage','structuredClone','queueMicrotask','Infinity',
 'NaN','undefined','null','true','false','this','super','arguments','globalThis','Boolean','Symbol',
}
keywords = {
 'const','let','var','function','return','if','else','for','while','do','switch','case','break',
 'continue','new','typeof','instanceof','delete','void','in','of','try','catch','finally','throw',
 'class','extends','static','async','await','yield','export','import','default','from','as',
}

problems = []
for m in re.finditer(r'\nfunction ([A-Z]\w*)\(\{([^)]*?)\}\)\s*\{', src, re.S):
    name, sig = m.group(1), m.group(2)
    start = m.end()
    nxt = src.find('\nfunction ', start)
    body = src[start:nxt if nxt > 0 else len(src)]
    body = strip_noise(body)
    # Only the JS above the first return — hooks, memos, derived values. Below
    # that is JSX, where prose and attribute values live.
    # Cut at the component's own JSX return (two-space indent), not at any
    # `return` inside a nested callback — those truncate the scan too early.
    m2 = re.search(r'\n  return \(', body)
    if not m2: continue
    js = body[:m2.start()]
    # Early returns can contain JSX (`if (!x) return <div />`); strip tags so
    # attribute names aren't read as variables.
    # Must only match real tags. The previous pattern took any `<` to the next
    # `>`, so `rows.length < 2` followed later by an arrow `(v) =>` deleted
    # everything between them — including three `const` declarations, which
    # were then reported as undeclared. Requiring a tag name and forbidding
    # newlines keeps it to actual JSX.
    # Requiring a tag name after `<` is what matters: `rows.length < 2` is not
    # a tag, so the scan no longer runs from a comparison to a later arrow and
    # deletes the declarations in between. Newlines stay allowed because real
    # JSX tags wrap across lines.
    js = re.sub(r'</?[A-Za-z][^<>]*>', ' ', js)

    props = set(re.findall(r'(\w+)\s*(?:=[^,]*)?(?:,|$)', sig))
    props |= set(re.findall(r'\b(\w+)\b', sig))
    declared = set(re.findall(r'(?:const|let|var|function)\s+(\w+)', js))
    # Comma-separated declarators: `const a = 1, b = 2;` declares both.
    for stmt in re.findall(r'(?:const|let|var)\s+([^;\n]+)', js):
        for part in stmt.split(','):
            nm = part.split('=')[0].strip()
            if re.match(r'^\w+$', nm): declared.add(nm)
    for a, b in re.findall(r'(?:const|let|var)\s*\[\s*(\w+)\s*,\s*(\w+)\s*\]', js):
        declared |= {a, b}
    for grp in re.findall(r'(?:const|let|var)\s*\{([^}]*)\}\s*=', js):
        declared |= {p.split(':')[-1].split('=')[0].strip() for p in grp.split(',')}
    # parameters of inner arrow functions and callbacks
    for grp in re.findall(r'\(([^()]*)\)\s*=>', js):
        declared |= {p.split('=')[0].replace('{','').replace('}','').replace('[','').replace(']','').strip()
                     for p in grp.split(',')}
    for one in re.findall(r'(\w+)\s*=>', js):
        declared.add(one)
    declared |= set(re.findall(r'catch\s*\(\s*(\w+)', js))

    used = set(re.findall(r'(?<![.\w$\'"`])([a-z_]\w*)\s*(?=[.\[)(,;=<>!?+\-*/&|\s])', js))
    unknown = sorted(u for u in used
                     if u not in props and u not in declared and u not in toplevel
                     and u not in builtins and u not in keywords)
    if unknown:
        line = src[:start].count('\n') + 1
        problems.append(f'{name} (line {line}): {", ".join(unknown)}')

# Verified false positives: destructuring patterns and JSX text this simple
# parser cannot model. Anything NOT on this list fails the run, so a genuinely
# orphaned reference — the bug that has produced three blank tabs — is caught
# before it ships rather than being lost in advisory noise.
KNOWN_OK = {
    "CompletionCalendar": {"a", "c", "start", "weeks"},
    "DeleteButton": {"className", "e", "style"},
    "ZoomableLineChart": {"onTouchMove", "r", "values"},
    "ParamHistoryModal": {"windowStats"},
    "TestLab": {"def", "key"},
    "LaunchSplash": {"b"},
    "CoralClump": {"b"},
    # Prose between JSX tags on the same line as an expression — the stripper
    #    reads "Cancel and go back to {fmtAmount(x)} mL/day" as identifiers. The
    #    words below are all plain English from that panel's copy.
    "CorrectionPanel": {"and", "back", "complete", "day", "dose", "go", "it",
                        "mL", "running", "to"},
    "ReminderSheet": {"intervalNum"},
    "LogResultPopup": {"t", "i", "timers"},
    "TaskDonePopup": {"a", "avg", "b", "t"},
    "IcpResultPopup": {"moves", "pct", "was"},
}

real = []
for line in problems:
    name = line.split(" (")[0]
    ids = {x.strip() for x in line.split(": ", 1)[1].split(",")}
    unknown = ids - KNOWN_OK.get(name, set())
    if unknown:
        real.append(f"{name}: {', '.join(sorted(unknown))}")

if real:
    print("FAIL identifiers used but never declared:")
    for r in real: print("   ", r)
    sys.exit(1)
print("OK   no new undeclared identifiers (%d known false positives allowed)"
      % sum(len(v) for v in KNOWN_OK.values()))
sys.exit(0)

# Advisory, not a gate. A precise version would need a real JS parser; this
# catches the common case (an orphaned reference left behind by an edit) at the
# cost of some false positives from destructuring it doesn't model. Worth
# scanning by eye after edits; the error boundary catches whatever slips past.
if problems:
    print('SUSPECTS (verify by eye — some are destructuring this cannot model):')
    for p in problems: print('   ', p)
else:
    print('OK   no undeclared identifiers in component bodies')

File: scripts/verify/scopecheck.mjs
#!/usr/bin/env node
/* Ported from legacy/tools/scopecheck.py.
 *
 * Finds identifiers used in a component's JS body — the hooks and derived
 * values above `return (` — that are never declared. This is what caught
 * "Can't find variable: editing" and a blank tab. Deliberately scoped to
 * components only, not engine functions: an undeclared read in an engine
 * throws the moment golden/protocols/run_all exercise it, so the behavioural
 * suites already guard that path. A component can go a long time unexercised
 * because nothing renders it.
 *
 * Change from the monolith: a file's own imports now count as declared, same
 * as the monolith's flat top-level scope counted every other top-level name
 * as declared regardless of where it sat in the file.
 */
import { listSrcFiles, readAll, importedNames } from './util.mjs';

const BUILTINS = new Set(['React', 'Math', 'Object', 'Array', 'Number', 'String', 'JSON', 'Date', 'Set',
  'Map', 'Promise', 'Error', 'console', 'window', 'document', 'navigator', 'fetch', 'setTimeout',
  'clearTimeout', 'setInterval', 'clearInterval', 'requestAnimationFrame', 'cancelAnimationFrame',
  'parseInt', 'parseFloat', 'isNaN', 'isFinite', 'Blob', 'FileReader', 'URL', 'encodeURIComponent',
  'decodeURIComponent', 'localStorage', 'sessionStorage', 'structuredClone', 'queueMicrotask', 'Infinity',
  'NaN', 'undefined', 'null', 'true', 'false', 'this', 'super', 'arguments', 'globalThis', 'Boolean',
  'Symbol']);
const KEYWORDS = new Set(['const', 'let', 'var', 'function', 'return', 'if', 'else', 'for', 'while', 'do',
  'switch', 'case', 'break', 'continue', 'new', 'typeof', 'instanceof', 'delete', 'void', 'in', 'of', 'try',
  'catch', 'finally', 'throw', 'class', 'extends', 'static', 'async', 'await', 'yield', 'export', 'import',
  'default', 'from', 'as']);

function stripNoise(t) {
  return t
    .replace(/\/\*[\s\S]*?\*\//g, ' ')
    .replace(/(?<=^|\n)\s*\/\/[^\n]*/g, ' ')
    .replace(/\/\/[^\n]*$/gm, ' ')
    .replace(/"(?:\\.|[^"\\])*"/g, '""')
    .replace(/'(?:\\.|[^'\\])*'/g, "''")
    .replace(/`(?:\\.|[^`\\])*`/g, '``');
}

// Verified false positives: destructuring patterns and JSX prose this simple
// parser cannot model. Ported verbatim from legacy scopecheck.py's KNOWN_OK,
// re-keyed to whichever file each component now lives in; unmatched entries
// are dropped below with a note, not carried as dead weight.
const KNOWN_OK = {
  CompletionCalendar: ['a', 'c', 'start', 'weeks'],
  DeleteButton: ['className', 'e', 'style'],
  ZoomableLineChart: ['onTouchMove', 'r', 'values'],
  ParamHistoryModal: ['windowStats'],
  TestLab: ['def', 'key'],
  LaunchSplash: ['b'],
  CoralClump: ['b'],
  CorrectionPanel: ['and', 'back', 'complete', 'day', 'dose', 'go', 'it', 'mL', 'running', 'to'],
  ReminderSheet: ['intervalNum'],
  LogResultPopup: ['t', 'i', 'timers'],
  TaskDonePopup: ['a', 'avg', 'b', 't'],
  IcpResultPopup: ['moves', 'pct', 'was'],
};

const files = listSrcFiles();
const entries = readAll(files);
const problems = [];
const usedAllowlist = new Set();

for (const { rel, src } of entries) {
  const toplevel = new Set([...src.matchAll(/^(?:export\s+)?(?:async\s+)?(?:function|const|let|var|class)\s+(\w+)/gm)].map((m) => m[1]));
  const imported = importedNames(src);

  for (const m of src.matchAll(/\nexport function ([A-Z]\w*)\(\{([^)]*?)\}\)\s*\{/gs)) {
    const [, name, sig] = m;
    const start = m.index + m[0].length;
    const nxt = src.indexOf('\nexport function ', start);
    let body = src.slice(start, nxt > 0 ? nxt : src.length);
    body = stripNoise(body);
    const m2 = body.match(/\n  return \(/);
    if (!m2) continue;
    let js = body.slice(0, m2.index);
    js = js.replace(/<\/?[A-Za-z][^<>]*>/g, ' ');

    const props = new Set([...sig.matchAll(/\b(\w+)\b/g)].map((x) => x[1]));
    const declared = new Set([...js.matchAll(/(?:const|let|var|function)\s+(\w+)/g)].map((x) => x[1]));
    for (const stmt of js.matchAll(/(?:const|let|var)\s+([^;\n]+)/g)) {
      for (const part of stmt[1].split(',')) {
        const nm = part.split('=')[0].trim();
        if (/^\w+$/.test(nm)) declared.add(nm);
      }
    }
    for (const dm of js.matchAll(/(?:const|let|var)\s*\[\s*(\w+)\s*,\s*(\w+)\s*\]/g)) { declared.add(dm[1]); declared.add(dm[2]); }
    for (const dm of js.matchAll(/(?:const|let|var)\s*\{([^}]*)\}\s*=/g)) {
      for (const p of dm[1].split(',')) {
        const nm = p.split(':').pop().split('=')[0].trim();
        if (nm) declared.add(nm);
      }
    }
    for (const dm of js.matchAll(/\(([^()]*)\)\s*=>/g)) {
      for (const p of dm[1].split(',')) {
        const nm = p.split('=')[0].replace(/[{}[\]]/g, '').trim();
        if (nm) declared.add(nm);
      }
    }
    for (const dm of js.matchAll(/(\w+)\s*=>/g)) declared.add(dm[1]);
    for (const dm of js.matchAll(/catch\s*\(\s*(\w+)/g)) declared.add(dm[1]);

    const used = new Set([...js.matchAll(/(?<![.\w$'"`])([a-z_]\w*)\s*(?=[.[)(,;=<>!?+\-*/&|\s])/g)].map((x) => x[1]));
    const unknown = [...used].filter((u) => !props.has(u) && !declared.has(u) && !toplevel.has(u)
      && !BUILTINS.has(u) && !KEYWORDS.has(u) && !imported.has(u));

    const allowed = new Set(KNOWN_OK[name] || []);
    const real = unknown.filter((u) => !allowed.has(u));
    unknown.filter((u) => allowed.has(u)).forEach((u) => usedAllowlist.add(`${name}.${u}`));
    if (real.length) {
      const line = src.slice(0, start).split('\n').length;
      problems.push(`${rel}:${line} ${name}: ${real.sort().join(', ')}`);
    }
  }
}

if (problems.length) {
  console.log('FAIL identifiers used but never declared:');
  for (const p of problems) console.log('    ' + p);
  process.exit(1);
}
console.log(`OK   no undeclared identifiers in component bodies (${usedAllowlist.size} known false positives allowed)`);
